[FLINK-28519][network] Fix the bug that SortMergeResultPartitionReadScheduler may not read data sequentially

TanYuxin-tyx opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Currently, the SortMergeResultPartitionReadScheduler always gets all active subpartition readers and read at most one data region for them. It is common that some subpartitions are requested before others and their region indexes are ahead of others. 

If all region data of a subpartition can be read in one round, some subpartition readers will always be ahead of others which will cause random IO. 

This patch fixes this case by polling one subpartition reader at a time.


## Brief change log
  - *Fix the bug that SortMergeResultPartitionReadScheduler may not read data sequentially.*


## Verifying this change

This change is already covered by existing tests, such as *SortMergeResultPartitionReadSchedulerTest*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

reswqa commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java:
```diff
@@ -152,12 +160,34 @@ class SortMergeResultPartitionReadScheduler implements Runnable, BufferRecycler
+        } catch (Throwable throwable) {
+            // fail all pending subpartition readers immediately if any exception occurs
+            LOG.error("Failed to request buffers for data reading.", throwable);
+            failSubpartitionReaders(getAllReaders(), throwable);
```
If a failed reader not removed from `sortedBuffers` in time, It can only be removed after the next poll, but this failed reader will read data from disk int that case, which is an unnecessary extra overhead. IMO, we need clear `sortedReaders`  when `allocateBuffers` throw exception.

reswqa replied:
`release` and `releaseSubpartitionReader` should take care of `sortedReaders ` also. These two method will add reader to `failedReaders` set, I suggest add check to `getNextReaderToRead`, when the polled reader contains in `failedReaders`, we can ignore it and poll next until meet the reader not failed.

TanYuxin-tyx replied:
@reswqa Thanks for helping review. Good point. 

In order to clear the reader in `sortedBuffers` in time, I added a check whether the reader is in `failedReaders` in the method `addPreviousAndGetNextReader`. 
In addtion, when `allReaders` is empty, the `sortedBuffers` will be cleared in `removeFinishedAndFailedReaders`.

reswqa reviewed (commented):
@TanYuxin-tyx, thanks for your contribution, this bug-fix looks good to me overall. I left some comments, please take a look~

TanYuxin-tyx commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeResultPartitionReadScheduler.java:
```diff
@@ -283,13 +284,26 @@ private void mayNotifyReleased() {
 
-            return new PriorityQueue<>(allReaders);
+    @Nullable
+    private SortMergeSubpartitionReader getNextReaderToRead(
```
OK, renamed it.

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
> @TanYuxin-tyx, thanks for your contribution, this bug-fix looks good to me overall. I left some comments, please take a look~

@reswqa Thanks a lot for reviewing the PR, I have updated the code and fixed the comments.

reswqa reviewed (approved):
@TanYuxin-tyx Thanks for addressing comments so quickly, LGTM!
cc @wsry

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@flinkbot run azure

TanYuxin-tyx commented:
@flinkbot run azure

wsry commented:
@TanYuxin-tyx Thanks for the change and review from @reswqa . LGTM.
